lnwire/onion_error.go

```diff
@@ -1222,18 +1223,41 @@ func DecodeFailure(r io.Reader, pver uint32) (FailureMessage, error) {
 	// is a 2 byte length followed by the payload itself.
 	var failureLength uint16
 	if err := ReadElement(r, &failureLength); err != nil {
-		return nil, fmt.Errorf("unable to read error len: %v", err)
-	}
-	if failureLength > FailureMessageLength {
-		return nil, fmt.Errorf("failure message is too "+
-			"long: %v", failureLength)
+		return nil, fmt.Errorf("unable to read failure len: %w", err)
 	}
+
 	failureData := make([]byte, failureLength)
 	if _, err := io.ReadFull(r, failureData); err != nil {
 		return nil, fmt.Errorf("unable to full read payload of "+
-			"%v: %v", failureLength, err)
+			"%v: %w", failureLength, err)
+	}
+
+	// Read the padding.
+	var padLength uint16
+	if err := ReadElement(r, &padLength); err != nil {
+		return nil, fmt.Errorf("unable to read pad len: %w", err)
+	}
+
+	if _, err := io.CopyN(ioutil.Discard, r, int64(padLength)); err != nil {
+		return nil, fmt.Errorf("unable to read padding %w", err)
+	}
+
+	// Verify that we are at the end of the stream now.
+	scratch := make([]byte, 1)
+	_, err := r.Read(scratch)
+	if err != io.EOF {
+		return nil, fmt.Errorf("unexpected failure bytes")
+	}
+
+	// Check the total length.
+	totalLength := padLength + failureLength
```

Comment:
`padLength` and `failureLength` are uint16 and they can overflow and still make `totalLength != FailureMessageLength` return false

Reply:
with the check `totalLength < FailureMessageLength`, the overflow can make a now-valid failure reason fail if the check is true

Reply:
i realize now that it's not possible to overflow given the constraints on message size, still it would be nice to convert to uint32

Reply:
Didn't think of this, even if it may not be a problem. Fixed anyway as suggested.